### PR TITLE
Implement `IndexConfig`

### PR DESCRIPTION
**Pull request by guilload**
Pushing a rough implementation to have a basis for discussion.

### CODE REVIEW

File: quickwit-index-config/resources/tests/index_object/hdfs-logs.toml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+key = "value"
+
+[indexing_settings.resources]
+num_threads = 1
+heap_size = "1G"
+
+[schema] # or mapping? mappings? 
+default_search_fields = [ "body", "title" ]
+store_source = true
+demux_field = "tenant_id"
```
Comment (guilload) [inline quickwit-index-config/resources/tests/index_object/hdfs-logs.toml:23]: Do we want this in `indexing_settings`?

File: quickwit-index-config/src/index_object.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+#[derive(Serialize, Deserialize)]
+pub struct SourceConfig {
+    pub source_id: String,
+    pub source_type: String,
+    pub params: toml::Value,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct IndexObject {
```
Comment (guilload) [inline quickwit-index-config/src/index_object.rs:63]: Will be renamed to `IndexConfig`... but first, we need to find a better name for the current `IndexConfig`.

File: quickwit-index-config/src/index_object.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+use crate::IndexConfig;
+
+#[derive(Serialize, Deserialize)]
+pub struct IndexingResources {
+    pub num_threads: usize,
+    pub heap_size: Byte,
+}
+
+// Is here a good centralized place to define those objects?
```
Comment (guilload) [inline quickwit-index-config/src/index_object.rs:33]: We can import those objects from `quickwit-indexing` but then `quickwit-index-config` depends on `quickwit-indexing`, and `quickwit-indexing` depends on `quickwit-index-config`. With that, `quickwit-metastore` now depends transitively on `quickwit-indexing`. Not great, from a dependency graph perspective. The compiler can figure this out though.

File: quickwit-index-config/src/index_object.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+pub struct IndexingSettings {
+    pub commit_policy: CommitPolicy,
+    pub demux_policy: DemuxPolicy,
+    pub merge_policy: MergePolicy,
+    pub retention_policy: RetentionPolicy,
+    pub resources: IndexingResources,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SourceConfig {
```
Comment (guilload) [inline quickwit-index-config/src/index_object.rs:56]: Duplicate from the struct in `quickwit-indexing`, will fix when we make a decision on "quickwit-models".

File: quickwit-index-config/src/index_object.rs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+pub struct IndexObject {
+    pub index_id: String,
+    pub index_uri: String,
+    // pub metastore_uri: String,
+    // pub index_mapping: Box<dyn IndexConfig>,
+    pub indexing_settings: IndexingSettings,
+    pub sources: Vec<SourceConfig>,
+}
+
+impl IndexObject {
```
Comment (guilload) [inline quickwit-index-config/src/index_object.rs:72]: Shall I add a `from_file` function that tries to infer the format of a file according to its extension?

File: quickwit-index-config/resources/tests/index_object/hdfs-logs.toml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+index_id = "hdfs-logs"
```
Comment (guilload) [inline quickwit-index-config/resources/tests/index_object/hdfs-logs.toml:1]: Correctness aside, at the end of the day, TOML is my least favorite format to work with.

File: quickwit-index-config/resources/tests/index_object/hdfs-logs.toml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+index_uri = "s3://quickwit-indexes/hdfs-logs"
+
+[indexing_settings.commit_policy]
```
Comment (guilload) [inline quickwit-index-config/resources/tests/index_object/hdfs-logs.toml:4]: Will complete the policy objects.

**Comment by guilload** [comment]
Feel free to drop some commits on this PR.

File: quickwit-index-config/resources/tests/index_object/hdfs-logs.toml (hunk 6 above)
Comment (fulmicoton) [reply quickwit-index-config/resources/tests/index_object/hdfs-logs.toml:1]: haha really?
Should we switch to yaml then?

File: quickwit-index-config/resources/tests/index_object/hdfs-logs.toml (hunk 1 above)
Comment (fulmicoton) [reply quickwit-index-config/resources/tests/index_object/hdfs-logs.toml:21]: Question for @fmassot... i am not sure of the implication...

File: quickwit-index-config/src/index_object.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    pub index_uri: String,
+    pub indexing_settings: IndexingSettings,
+    pub sources: Vec<SourceConfig>,
+}
+
+impl IndexObject {
+    // TODO: asyncify?
+    pub fn from_file<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+        let parser_fn = match path.as_ref().extension().and_then(OsStr::to_str) {
+            Some("json") => Self::from_json,
```
Comment (fulmicoton) [inline quickwit-index-config/src/index_object.rs:73]: Ah ok so we handle everything. Cool.

File: quickwit-index-config/resources/tests/index_object/hdfs-logs.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+            "merge_factor": 10,
+            "max_merge_factor": 12,
+            "min_level_num_docs": 100000
+        },
+        "resources": {
+            "num_threads": 1,
+            "heap_size": "1G"
+        }
+    },
+    "sources": [
```
Comment (fulmicoton) [inline quickwit-index-config/resources/tests/index_object/hdfs-logs.json:18]: What's today in indexconfig (schema stuff most notably) will end up nest in this object too or not?

Comment (guilload) [reply quickwit-index-config/resources/tests/index_object/hdfs-logs.json:18]: In the current spec in Notion, `index_mapping` is at the root of the object. I see what you're getting at though. Depending on the source, docs could be different, but we would like to map them to the same index schema eventually.

We could define an index schema at the root and, in each source, a set of mappings/transformations (pseudo-DSL): 

```yaml
[...]
index_schema:
  - ts:
    type: u64
  - title:
    type: text
  - body:
    type: text

sources:
  - source_id: hdfs-logs-kafka-source
    source_type: kafka
    transforms:
      - rename:
        source: content
        target: body
      - drop:
        source: url
    params:
      bootstrap.servers: host:9092
      topic: cloudera-cluster-logs
    
  - source_id: hdfs-logs-kinesis-source
    source_type: kinesis
    transforms:
      - rename:
        source: timestamp
        target: ts
      - drop:
        source: url
    params:
      stream_name: emr-cluster-logs
```

We'll have to put more thought into it.

**Review by fulmicoton — APPROVED** [review]

File: quickwit-config/src/index_config.rs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct SearchSettings {
+    pub default_search_fields: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SourceConfig {
+    pub source_id: String,
+    pub source_type: String,
+    pub params: toml::Value,
```
Comment (fmassot) [inline quickwit-config/src/index_config.rs:69]: what are the advantages of using `toml::Value` over a json value?

Comment (guilload) [reply quickwit-config/src/index_config.rs:69]: Limited, primarily support for date and time types. I thought toml would be a first-class citizen... maybe not :)

**Pull request merged**